File: clojure/aoc/src/aoc/y21/day18.clj
(ns aoc.y21.day18
  (:require [aoc.utils :as u]))

; sfish-reduce
(defn sfish-reduce [num]
  (println "Implement me"))

(defn explode [num])

(defn sfish-add [l r]
  [l r]
  )

(defn split
  "To split a regular number, replace it with a pair; the left element of the pair should be the regular number divided by two and rounded down, while the right element of the pair should be the regular number divided by two and rounded up. For example, 10 becomes [5,5], 11 becomes [5,6], 12 becomes [6,6], and so on."
  [num]
  (assert (number? num))
  [(int (Math/floor (/ num 2)))
   (int (Math/ceil (/ num 2)))])

(defn magnitude
  "The magnitude of a pair is 3 times the magnitude of its left element plus 2 times the magnitude of its right element. The magnitude of a regular number is just that number."
  [snailfish-num]
  (cond
    (number? snailfish-num) snailfish-num
    :else (+ 
           (* 3 (magnitude (first snailfish-num))) 
           (* 2 (magnitude (second snailfish-num))))))

File: python/aoc/y21/day_18.py
import math

def read_input():
    with open('data/day18.txt', 'r') as f:
        lines = f.readlines()
        return [eval(line.strip()) for line in lines]

def add(left, right):
    return [left, right]


def reduce(num):
    # if any pair is nested inside four pairs, the leftmost such pair explodes.

    # If any regular number is 10 or greater, the leftmost such regular number splits.

    pass


def explode(pair):
    assert len(pair == 2), f'len({pair}) = {len(pair)}. Expected 2'
    left = pair[0]
    right = pair[1]
    # add pair's left value to first regular number to left of exploding pair
    # add pair's right value to first regular number to right of exploding pair
    #

    pass

def split(i):
    return [math.floor(i / 2), math.ceil(i / 2)]

def magnitude(snailfish_num):
    if isinstance(snailfish_num, int):
        return snailfish_num
    return 3 * magnitude(snailfish_num[0]) + 2 * magnitude(snailfish_num[1])

if __name__ == '__main__':
    snailfish_num_list = read_input()
    reduce(eval('[[[[[9,8],1],2],3],4]'))
